### PR TITLE
[HttpFoundation] Use `Cache-Control: must-revalidate` only if explicit lifetime has been given

### PR DESCRIPTION
**Pull request by mpdude**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | yes
| New feature?  | no
| Deprecations? | no 
| Tickets       | 
| License       | MIT
| Doc PR        | 

This is really nit-picking: The conservative, safe default for `Cache-Control` is `private, no-cache` which means the response must not be served from cache unless it has been validated.

If `Last-Modified` or `Expires` are present, we can relax `no-cache` to be `must-revalidate`, which means that _once the response has become stale_, it must be revalidated.

An `ETag` alone does not give the response a lifetime, so IMO sticking with `no-cache` in this case would be more consistent.

### CODE REVIEW

**Comment by derrabus** [comment]
> An `ETag` alone does not give the response a lifetime

`Last-Modified` doesn't either, does it?

**Comment by mpdude** [comment]
TIL that if the status code is [cacheable by default](https://tools.ietf.org/html/rfc7231#section-6.1) or a response without explicit freshness has been marked as explicitly cacheable (e.g., with a "public"  response directive), then [heuristic freshness](https://tools.ietf.org/html/rfc7234#section-4.2.2) may be applied.

Until now, I always thought this was a browser quirk.

**Comment by nicolas-grekas** [comment]
Should the PR target 3.4?

**Comment by nicolas-grekas** [comment]
I don't understand the change. When there is an ETag, we should send `private, must-revalidate`, because that's what an ETag allows: revalidation.

**Comment by mpdude** [comment]
`no-cache` is the default, which tells a cache that it *must revalidate* the resource upon *every* request. `must-revalidate` means that revalidation must occur *once the response has become stale*.

As an `ETag` alone does not define a lifetime, my suggestion was to switch from `no-cache` to `must-revalidate` only if a header is present that defines the lifetime (as `Expires` or `Cache-Control: max-age=...`) or that allows for a heuristic expiration to be applied (`Last-Modified` is provided).

Nit-picking, really.

**Comment by mpdude** [comment]
I agree the RFC might have chosen more intention-revealing names.

**Comment by nicolas-grekas** [comment]
An ETag without a lifetime is considered as having a zero lifetime by browsers AFAIK.
I don't think no-cache is appropriate: a conditional request is possible today and works, whereas this change would kill this conditional request.

**Comment by mpdude** [comment]
No, it would not.

If the lifetime is considered to be zero, `no-cache` and `must-revalidate` would have the same effect, namely that a revalidation has to happen right from the start/after 0 seconds.

`no-store`: Prohibits caching, thus no revalidation/conditional requests.

`no-cache`: Response can be cached, but must be revalidated every time.

`must-revalidate`: Response may be used without validation as long as it is fresh; must be revalidated afterwards and must not be used when stale.

**Comment by nicolas-grekas** [comment]
Oh, so no-cache <=> must-revalite,maxage=0 got it. Works then.

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Tests do not pass apparently.

**Comment by mpdude** [comment]
Failing tests come from `ResponseCacheStrategy` ([here](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php)).

With the change from this PR, we now have `Cache-Control: no-cache` in some cases where previously we had `Cache-Control: must-revalidate`.

If this change applies to an ESI subrequest, `ResponseCacheStrategy` now decides ([here](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php#L163-L167)) that the resulting, ESI-tags-merged response can no longer be cached using the expiration model ([here](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpKernel/HttpCache/ResponseCacheStrategy.php#L112-L122)).

It thus adds an `Expires` header and sets it to the response `Date`. This is where the tests break: Now the lifetime is `0`, previously it was `null`.

I need to think through the `ResponseCacheStrategy` behaviour because it *feels* as if there might be a bug. The change here should not make a difference for the decision if something is cacheable or not.

Second, I'll try to remove the `Expires` header which IMO is not really necessary; not having it might already make tests pass again.

TL;DR: Working on it ;-)

**Comment by mpdude** [comment]
Maybe @nicolas-grekas can confirm that the `deps=high` build failure is due to the change being in one component, the fix for the test in another one?

**Review by nicolas-grekas — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
Thank you @mpdude.
